File: Mapgenerator/Mapgenerator.py
from dataclasses import dataclass
from random import sample, random

@dataclass
class Point:
    x: int
    y: int
    z: int

    def __hash__(self):
        return hash(f"({self.x}, {self.y}, {self.z}")

@dataclass
class BrickDef:
    width: int
    depth: int
    color: str
    points: frozenset[Point]

    def shares_no_points(self, other: frozenset[Point]) -> bool:
        return self.points.isdisjoint(other)

@dataclass
class Config:
    nr_bricks: int
    max_height: int = None
    min_height: int = None


def find_placeable_spots(brick: BrickDef, 
                        placed_bricks: list[BrickDef], 
                        available_pegs: set[Point]) -> None | list[frozenset[Point]]:
    possible_points = []

    orientations = [(brick.width, brick.depth), (brick.depth, brick.width)]
    
    for peg in available_pegs:
        for width, depth in orientations:
            for xOffset in range(width):
                for zOffset in range(depth):
                    coordinates = frozenset(
                        Point(peg.x - xOffset + x, peg.y, peg.z - zOffset + z)
                        for x in range(width)
                        for z in range(depth)
                    )
                    if all(brick.shares_no_points(coordinates) for brick in placed_bricks):
                        possible_points.append(coordinates)

    return possible_points


def get_max_height(placed_bricks: list[BrickDef]) -> int:
    """Get the maximum height achieved in the placed bricks."""
    if not placed_bricks:
        return 0
    return max(point.y for brick in placed_bricks for point in brick.points)


def select_optimal_spot(spots: list[frozenset[Point]], definition: Config, placed_bricks: list[BrickDef], bricks_remaining: int) -> frozenset[Point]:
    """
    Select the optimal spot for placing a brick.
    If min_height is specified, scale aggressiveness based on how close we are to running out of safety margin.
    """
    if definition.min_height is None:
        return sample(spots, 1)[0]

    current_max_height = get_max_height(placed_bricks)

    # If we've already reached minimum height, play naturally with slight height preference
    if current_max_height >= definition.min_height:
        max_height_in_spots = max(min(point.y for point in spot) for spot in spots)
        max_height_spots = [spot for spot in spots if min(point.y for point in spot) == max_height_in_spots]

        # Mild preference for higher spots (60% vs 40%) for continued building
        if len(max_height_spots) > 0 and random() < 0.6:
            return sample(max_height_spots, 1)[0]
        else:
            return sample(spots, 1)[0]

    # We haven't reached minimum height yet - calculate aggressiveness based on safety margin
    height_deficit = definition.min_height - current_max_height
    safety_margin = bricks_remaining - height_deficit

    # Scale aggressiveness: more remaining pieces = less aggressive
    if safety_margin >= 3:
        # Plenty of safety margin - play almost completely naturally
        aggressiveness = 0.1  # 10% bias toward height
    elif safety_margin == 2:
        # Some safety margin - mild bias
        aggressiveness = 0.3  # 30% bias toward height
    elif safety_margin == 1:
        # Low safety margin - moderate bias
        aggressiveness = 0.6  # 60% bias toward height
    elif safety_margin == 0:
        # No safety margin - strong bias but not 100%
        aggressiveness = 0.8  # 80% bias toward height
    else:
        # Negative safety margin - MUST be aggressive
        aggressiveness = 1.0  # 100% bias toward height

    max_height_in_spots = max(min(point.y for point in spot) for spot in spots)
    max_height_spots = [spot for spot in spots if min(point.y for point in spot) == max_height_in_spots]

    # Apply scaled aggressiveness
    if len(max_height_spots) > 0 and random() < aggressiveness:
        # When being aggressive, prefer spots that build on existing structures for better stacking
        if aggressiveness >= 0.8 and len(placed_bricks) > 0:
            building_on_existing = []
            for spot in max_height_spots:
                for point in spot:
                    point_below = Point(point.x, point.y - 1, point.z)
                    if any(point_below in brick.points for brick in placed_bricks):
                        building_on_existing.append(spot)
                        break
            if building_on_existing:
                return sample(building_on_existing, 1)[0]

        return sample(max_height_spots, 1)[0]
    else:
        return sample(spots, 1)[0]


def generate_map(definition: Config) -> list[BrickDef]:
    # Basic validation - need at least enough bricks to build a tower
    # More conservative estimate: need at least min_height bricks to guarantee reaching min_height
    if definition.min_height is not None and definition.nr_bricks < definition.min_height:
        raise Exception(f"Cannot achieve minimum height {definition.min_height} with only {definition.nr_bricks} bricks")

    # Try to generate a map multiple times if min_height requirement isn't met
    max_attempts = 10 if definition.min_height is not None else 1

    for attempt in range(max_attempts):
        try:
            placed_bricks = generate_single_map(definition)

            # Validate that minimum height was achieved
            if definition.min_height is not None:
                max_height_achieved = get_max_height(placed_bricks)
                if max_height_achieved < definition.min_height:
                    if attempt == max_attempts - 1:  # Last attempt
                        raise Exception(f"Failed to achieve minimum height {definition.min_height} after {max_attempts} attempts. Only reached height {max_height_achieved}")
                    continue  # Try again

            Assert_no_overlapping_bricks(placed_bricks)
            return placed_bricks

        except Exception as e:
            if attempt == max_attempts - 1:  # Last attempt
                raise e
            continue  # Try again

    # This should never be reached, but just in case
    raise Exception("Failed to generate map")


def generate_single_map(definition: Config) -> list[BrickDef]:
    baseplate = BrickDef(6, 6, "gray", frozenset(Point(x, 0, z) for x in range(6) for z in range(6)))

    available_pegs: set[Point] = set(baseplate.points)
    available_bricks = get_available_bricks(definition.nr_bricks)

    placed_bricks: list[BrickDef] = []
    for i, brick in enumerate(available_bricks):
        bricks_remaining = len(available_bricks) - i - 1  # How many bricks left after this one

        spots = find_placeable_spots(brick, placed_bricks, available_pegs)
        if not spots:
            raise Exception("No spots available")

        spot = select_optimal_spot(spots, definition, placed_bricks, bricks_remaining)
        placed_bricks.append(BrickDef(brick.width, brick.depth, brick.color, spot))
        available_pegs.difference_update(spot)
        add_new_available_pegs(available_pegs, definition, placed_bricks, spots, spot)

    return placed_bricks

def Assert_no_overlapping_bricks(placed_bricks):
    for i in range(len(placed_bricks)-1):
        for j in range(i+1, len(placed_bricks)):
            assert(placed_bricks[i].points.isdisjoint(placed_bricks[j].points))

def add_new_available_pegs(available_pegs, definition: Config, placed_bricks, spots, spot):
    for p in spot:
        if p.y+1 >= definition.max_height:
            # Not allowed, too high.
            continue
        abovepoint = Point(p.x, p.y + 1, p.z)
        if all(brick.shares_no_points(frozenset([abovepoint])) for brick in placed_bricks):
            available_pegs.add(abovepoint)

        if p.y <= 0:  # Only add hanging points above baseplate level
            continue
        hanging_point = Point(p.x, p.y - 1, p.z)
        if all(brick.shares_no_points(frozenset([hanging_point])) for brick in placed_bricks):
            available_pegs.add(hanging_point)

def get_available_bricks(nr_bricks: int) -> list[BrickDef]:
    available_bricks = []
    for color in ["Yellow", "Red", "Green", "Blue"]:
        for i in range(1, 3):
            for j in range(i, 5):
                available_bricks.append(BrickDef(i, j, color, frozenset()))
    return sample(available_bricks, nr_bricks)

File: static/js/MapGenerator.js
// Port of Python Mapgenerator.py to JavaScript

export class Point {
    constructor(x, y, z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    toString() {
        return `(${this.x}, ${this.y}, ${this.z})`;
    }

    equals(other) {
        return this.x === other.x && this.y === other.y && this.z === other.z;
    }

    static fromArray(arr) {
        return new Point(arr[0], arr[1], arr[2]);
    }
}

export class BrickDef {
    constructor(width, depth, color, points) {
        this.width = width;
        this.depth = depth;
        this.color = color;
        this.points = new Set(points); // Set of Point objects
    }

    sharesNoPoints(otherPoints) {
        for (const point of this.points) {
            for (const otherPoint of otherPoints) {
                if (point.equals(otherPoint)) {
                    return false;
                }
            }
        }
        return true;
    }
}

export class Config {
    constructor(nrBricks, maxHeight = null, minHeight = null, excludeMiddle = false) {
        this.nrBricks = nrBricks;
        this.maxHeight = maxHeight;
        this.minHeight = minHeight;
        this.excludeMiddle = excludeMiddle; // Exclude middle 2x2 pegs from bottom plate
    }
}

function findPlaceableSpots(brick, placedBricks, availablePegs, config) {
    const possiblePoints = [];
    const possiblePointsMetadata = []; // Track which spots overlap middle 2x2
    const orientations = [
        [brick.width, brick.depth],
        [brick.depth, brick.width]
    ];

    for (const peg of availablePegs) {
        for (const [width, depth] of orientations) {
            for (let xOffset = 0; xOffset < width; xOffset++) {
                for (let zOffset = 0; zOffset < depth; zOffset++) {
                    const coordinates = [];
                    for (let x = 0; x < width; x++) {
                        for (let z = 0; z < depth; z++) {
                            coordinates.push(
                                new Point(peg.x - xOffset + x, peg.y, peg.z - zOffset + z)
                            );
                        }
                    }

                    // Check if all placed bricks share no points with these coordinates
                    if (placedBricks.every(b => b.sharesNoPoints(coordinates))) {
                        possiblePoints.push(coordinates);

                        // Track if this spot overlaps with middle 2x2 (for weighting later)
                        if (config.excludeMiddle) {
                            const overlapsMiddle = coordinates.some(point =>
                                point.y === 0 &&
                                point.x >= 2 && point.x <= 3 &&
                                point.z >= 2 && point.z <= 3
                            );
                            possiblePointsMetadata.push({ overlapsMiddle });
                        } else {
                            possiblePointsMetadata.push({ overlapsMiddle: false });
                        }
                    }
                }
            }
        }
    }

    return { spots: possiblePoints, metadata: possiblePointsMetadata };
}

function getMaxHeight(placedBricks) {
    if (placedBricks.length === 0) return 0;
    let maxHeight = 0;
    for (const brick of placedBricks) {
        for (const point of brick.points) {
            if (point.y > maxHeight) {
                maxHeight = point.y;
            }
        }
    }
    return maxHeight;
}

/**
 * Weighted random selection - picks an item based on weights
 * @param {Array} items - Array of items to choose from
 * @param {Array} weights - Array of weights (same length as items)
 * @param {Function} randomFn - Random function to use
 * @returns Selected item
 */
function weightedRandomSelect(items, weights, randomFn) {
    const totalWeight = weights.reduce((sum, w) => sum + w, 0);
    let random = randomFn() * totalWeight;

    for (let i = 0; i < items.length; i++) {
        random -= weights[i];
        if (random <= 0) {
            return items[i];
        }
    }

    // Fallback (shouldn't happen)
    return items[items.length - 1];
}

function selectOptimalSpot(spots, metadata, config, placedBricks, bricksRemaining, randomFn) {
    // Calculate weights for each spot based on its minimum height
    const spotHeights = spots.map(spot => Math.min(...spot.map(p => p.y)));

    // Determine height preference multiplier based on situation
    let heightMultiplier;

    if (config.minHeight === null) {
        // No minimum height requirement: mild preference for height (more natural building)
        heightMultiplier = 1.5;
    } else {
        const currentMaxHeight = getMaxHeight(placedBricks);

        if (currentMaxHeight >= config.minHeight) {
            // Already reached minimum height: mild preference for continued upward building
            heightMultiplier = 1.5;
        } else {
            // Haven't reached minimum height yet
            const heightDeficit = config.minHeight - currentMaxHeight;
            const safetyMargin = bricksRemaining - heightDeficit;

            if (safetyMargin <= 0) {
                // Critical: MUST reach height - extremely aggressive
                heightMultiplier = 100.0;
            } else if (safetyMargin === 1) {
                // Very low safety: very aggressive
                heightMultiplier = 20.0;
            } else if (safetyMargin === 2) {
                // Some safety: moderate-strong preference
                heightMultiplier = 2.5;
            } else {
                // Good safety: moderate preference
                heightMultiplier = 2.0;
            }
        }
    }

    // Calculate weights for each spot: weight = heightMultiplier ^ height
    // This gives exponentially more weight to higher spots, but fairly distributed
    const weights = spotHeights.map(height => Math.pow(heightMultiplier, height));

    // Apply penalty for spots that overlap with middle 2x2 (10% of normal weight = 90% chance to avoid)
    for (let i = 0; i < spots.length; i++) {
        if (metadata[i].overlapsMiddle) {
            weights[i] *= 0.1; // Strong disincentive (90% chance to place elsewhere)
        }
    }

    // Use weighted random selection
    return weightedRandomSelect(spots, weights, randomFn);
}

function assertNoOverlappingBricks(placedBricks) {
    for (let i = 0; i < placedBricks.length - 1; i++) {
        for (let j = i + 1; j < placedBricks.length; j++) {
            if (!placedBricks[i].sharesNoPoints(Array.from(placedBricks[j].points))) {
                throw new Error("Overlapping bricks detected!");
            }
        }
    }
}

function addNewAvailablePegs(availablePegs, config, placedBricks, spot, random) {
    for (const p of spot) {
        // Add peg above if within max height limit (or no limit)
        if (config.maxHeight === null || p.y + 1 < config.maxHeight) {
            const abovePoint = new Point(p.x, p.y + 1, p.z);
            if (placedBricks.every(brick =>
                brick.sharesNoPoints([abovePoint])
            )) {
                // Add to set using string representation to avoid duplicates
                availablePegs.add(abovePoint);
            }
        }

        // Add hanging peg below (only above baseplate level)
        if (p.y > 0) {
            const hangingPoint = new Point(p.x, p.y - 1, p.z);
            if (placedBricks.every(brick =>
                brick.sharesNoPoints([hangingPoint])
            )) {
                availablePegs.add(hangingPoint);
            }
        }
    }
}

function getAvailableBricks(nrBricks, seed = null) {
    const availableBricks = [];
    const colors = ["Yellow", "Red", "Green", "Blue"];

    for (const color of colors) {
        for (let i = 1; i <= 2; i++) {
            for (let j = i; j < 5; j++) {
                availableBricks.push(new BrickDef(i, j, color, []));
            }
        }
    }

    // If seed is provided, use it for deterministic shuffling
    if (seed !== null) {
        shuffleWithSeed(availableBricks, seed);
    } else {
        // Fisher-Yates shuffle
        for (let i = availableBricks.length - 1; i > 0; i--) {
            const j = Math.floor(Math.random() * (i + 1));
            [availableBricks[i], availableBricks[j]] = [availableBricks[j], availableBricks[i]];
        }
    }

    return availableBricks.slice(0, nrBricks);
}

// Seeded random number generator (LCG)
function seededRandom(seed) {
    let state = seed;
    return function() {
        state = (state * 1664525 + 1013904223) % 4294967296;
        return state / 4294967296;
    };
}

function shuffleWithSeed(array, seed) {
    const random = seededRandom(seed);
    for (let i = array.length - 1; i > 0; i--) {
        const j = Math.floor(random() * (i + 1));
        [array[i], array[j]] = [array[j], array[i]];
    }
}

function generateSingleMap(config, seed = null) {
    // Create baseplate
    const baseplatePoints = [];
    for (let x = 0; x < 6; x++) {
        for (let z = 0; z < 6; z++) {
            baseplatePoints.push(new Point(x, 0, z));
        }
    }
    const baseplate = new BrickDef(6, 6, "gray", baseplatePoints);

    const availablePegs = new Set(baseplatePoints);
    const availableBricks = getAvailableBricks(config.nrBricks, seed);

    const placedBricks = [];

    // Use seeded random if provided
    const random = seed !== null ? seededRandom(seed + 1000) : Math.random;

    for (let i = 0; i < availableBricks.length; i++) {
        const brick = availableBricks[i];
        const bricksRemaining = availableBricks.length - i - 1;

        const { spots, metadata } = findPlaceableSpots(brick, placedBricks, availablePegs, config);
        if (spots.length === 0) {
            throw new Error("No spots available");
        }

        // Always use optimal spot selection (respects min height requirements)
        const spot = selectOptimalSpot(spots, metadata, config, placedBricks, bricksRemaining, random);

        placedBricks.push(new BrickDef(brick.width, brick.depth, brick.color, spot));

        // Remove used points from available pegs
        for (const point of spot) {
            availablePegs.delete(point);
        }

        addNewAvailablePegs(availablePegs, config, placedBricks, spot, random);
    }

    return placedBricks;
}

export function generateMap(config, seed = null) {
    // Basic validation
    if (config.minHeight !== null && config.nrBricks < config.minHeight) {
        throw new Error(
            `Cannot achieve minimum height ${config.minHeight} with only ${config.nrBricks} bricks`
        );
    }

    const maxAttempts = config.minHeight !== null ? 10 : 1;

    for (let attempt = 0; attempt < maxAttempts; attempt++) {
        try {
            const placedBricks = generateSingleMap(config, seed);

            // Validate minimum height
            if (config.minHeight !== null) {
                const maxHeightAchieved = getMaxHeight(placedBricks);
                if (maxHeightAchieved < config.minHeight) {
                    if (attempt === maxAttempts - 1) {
                        throw new Error(
                            `Failed to achieve minimum height ${config.minHeight} after ${maxAttempts} attempts. Only reached height ${maxHeightAchieved}`
                        );
                    }
                    continue;
                }
            }

            assertNoOverlappingBricks(placedBricks);
            return placedBricks;
        } catch (e) {
            if (attempt === maxAttempts - 1) {
                throw e;
            }
            continue;
        }
    }

    throw new Error("Failed to generate map");
}

// Convert BrickDef array to BrickMap format compatible with existing editor
export function brickDefsToMapData(brickDefs) {
    const bricks = brickDefs.map(brickDef => {
        const points = Array.from(brickDef.points).map(p => ({
            x: p.x,
            y: p.y,
            z: p.z
        }));

        const xs = points.map(p => p.x);
        const zs = points.map(p => p.z);
        const width = Math.abs(Math.min(...xs) - Math.max(...xs)) + 1;
        const depth = Math.abs(Math.min(...zs) - Math.max(...zs)) + 1;
        const [finalWidth, finalDepth] = [Math.min(width, depth), Math.max(width, depth)];

        return {
            color: brickDef.color,
            name: `${finalWidth}x${finalDepth} ${brickDef.color}`,
            points: points
        };
    });

    return {
        metadata: {
            width: 6,
            height: 1,
            depth: 6,
            name: "Generated Map",
            timestamp: new Date().toISOString()
        },
        bricks: bricks
    };
}
